subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c

```diff
@@ -0,0 +1,1295 @@
+#define PGPS_REQ_PREDICT_INT_MIN	"&" NRF_CLOUD_JSON_PGPS_INT_MIN "=%u"
+#define PGPS_REQ_START_GPS_DAY		"&" NRF_CLOUD_JSON_PGPS_GPS_DAY "=%u"
+#define PGPS_REQ_START_GPS_TOD_S	"&" NRF_CLOUD_JSON_PGPS_GPS_TIME "=%u"
+
+#define HTTP_PROTOCOL "HTTP/1.1"
+#define SOCKET_PROTOCOL IPPROTO_TLS_1_2
+
+#define JITP_HOSTNAME_TLS	CONFIG_NRF_CLOUD_HOST_NAME
+#define JITP_PORT 		8443
+#define JITP_URL		"/topics/jitp?qos=1"
+#define JITP_CONTENT		"*/*\r\n" \
+				"Connection: close\r\n" \
+				"Host: " JITP_HOSTNAME_TLS ":" STRINGIFY(JITP_PORT) "\r\n"
+#define JITP_HTTP_TIMEOUT_MS 	(15000)
+#define JITP_RX_BUF_SZ		(255)
+
+/* Mapping of enum to strings for Job Execution Status. */
+static const char *const job_status_strings[] = {
+	[NRF_CLOUD_FOTA_QUEUED]      = "QUEUED",
+	[NRF_CLOUD_FOTA_IN_PROGRESS] = "IN_PROGRESS",
+	[NRF_CLOUD_FOTA_FAILED]      = "FAILED",
+	[NRF_CLOUD_FOTA_SUCCEEDED]   = "SUCCEEDED",
+	[NRF_CLOUD_FOTA_TIMED_OUT]   = "TIMED_OUT",
+	[NRF_CLOUD_FOTA_REJECTED]    = "REJECTED",
+	[NRF_CLOUD_FOTA_CANCELED]    = "CANCELLED",
+	[NRF_CLOUD_FOTA_DOWNLOADING] = "DOWNLOADING",
+};
+#define JOB_STATUS_STRING_COUNT (sizeof(job_status_strings) / \
+				 sizeof(*job_status_strings))
+
+/* Mapping of enum to strings for AGPS request type. */
+static const char *const agps_req_type_strings[] = {
+	[NRF_CLOUD_REST_AGPS_REQ_ASSISTANCE]	= AGPS_REQ_TYPE_STR_ASSIST,
+	[NRF_CLOUD_REST_AGPS_REQ_LOCATION]	= AGPS_REQ_TYPE_STR_LOC,
+	[NRF_CLOUD_REST_AGPS_REQ_CUSTOM]	= AGPS_REQ_TYPE_STR_CUSTOM,
+};
+
+static void http_response_cb(struct http_response *rsp,
+			enum http_final_call final_data,
+			void *user_data)
+{
+	struct nrf_cloud_rest_context *rest_ctx = NULL;
+
+	if (user_data) {
+		rest_ctx = (struct nrf_cloud_rest_context *)user_data;
+	}
+
+	if (rest_ctx && rsp->body_found && rsp->body_start) {
+		rest_ctx->response = rsp->body_start;
+	}
+
+	if (final_data == HTTP_DATA_FINAL) {
+		LOG_DBG("HTTP: All data received, status: %u %s",
+			rsp->http_status_code,
+			log_strdup(rsp->http_status));
+
+		if (!rest_ctx) {
+			LOG_WRN("User data not provided");
+			return;
+		}
+
+		rest_ctx->status = rsp->http_status_code;
+		rest_ctx->response_len = rsp->content_length;
+	}
+}
+
+static int generate_auth_header(const char *const tok, char **auth_hdr_out)
+{
+	if (!tok || !auth_hdr_out) {
+		return -EINVAL;
+	}
+
+	int ret;
+	size_t buff_size = sizeof(AUTH_HDR_BEARER_TEMPLATE) + strlen(tok);
+
+	*auth_hdr_out = k_calloc(buff_size, 1);
+	if (!*auth_hdr_out) {
+		return -ENOMEM;
+	}
+	ret = snprintk(*auth_hdr_out, buff_size, AUTH_HDR_BEARER_TEMPLATE, tok);
+	if (ret < 0 || ret >= buff_size) {
+		k_free(*auth_hdr_out);
+		*auth_hdr_out = NULL;
+		return -ETXTBSY;
+	}
+
+	return 0;
+}
+
+int tls_setup(int fd, const char *const tls_hostname, const sec_tag_t sec_tag)
+{
+	int err;
+	int verify;
+	const sec_tag_t tls_sec_tag[] = {
+		sec_tag,
+	};
+
+	enum {
+		NONE = 0,
+		OPTIONAL = 1,
+		REQUIRED = 2,
+	};
+
+	verify = REQUIRED;
```

Comment:
Peer verification flags are now defined in Zephyr, so they can be used instead:
https://github.com/nrfconnect/sdk-zephyr/blob/e29515afd99d13eb111ba209b10b722ead299b03/include/net/socket.h#L145-L148